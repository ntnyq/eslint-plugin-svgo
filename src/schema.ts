import type { JSONSchema4 } from 'json-schema'

/**
 * Stringify Options
 *
 * Bellow options are not supported
 * - `regEntities`
 * - `regValEntities`
 * - `encodeEntity`
 */
const js2svg = {
  type: 'object',
  properties: {
    doctypeStart: {
      type: 'string',
    },
    doctypeEnd: {
      type: 'string',
    },
    procInstStart: {
      type: 'string',
    },
    procInstEnd: {
      type: 'string',
    },
    tagOpenStart: {
      type: 'string',
    },
    tagOpenEnd: {
      type: 'string',
    },
    tagCloseStart: {
      type: 'string',
    },
    tagCloseEnd: {
      type: 'string',
    },
    tagShortStart: {
      type: 'string',
    },
    tagShortEnd: {
      type: 'string',
    },
    attrStart: {
      type: 'string',
    },
    attrEnd: {
      type: 'string',
    },
    commentStart: {
      type: 'string',
    },
    commentEnd: {
      type: 'string',
    },
    cdataStart: {
      type: 'string',
    },
    cdataEnd: {
      type: 'string',
    },
    textStart: {
      type: 'string',
    },
    textEnd: {
      type: 'string',
    },
    indent: {
      type: 'number',
    },
    pretty: {
      type: 'boolean',
    },
    useShortTags: {
      type: 'boolean',
    },
    eol: {
      type: 'string',
      enum: ['lf', 'crlf'],
    },
    finalNewline: {
      type: 'boolean',
    },
  },
} satisfies JSONSchema4

const plugins = {
  type: 'array',
  items: {
    type: 'string',
  },
} satisfies JSONSchema4

export const svgoConfigProperties = {
  path: {
    type: 'string',
  },
  multipass: {
    type: 'boolean',
  },
  floatPrecision: {
    type: 'integer',
    minimum: 0,
  },
  datauri: {
    type: 'string',
    enum: ['base64', 'enc', 'unenc'],
  },
  js2svg,
  plugins,
} satisfies JSONSchema4['properties']
