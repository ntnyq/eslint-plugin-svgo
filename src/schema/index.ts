import { PLUGIN_PRESET_DEFAULT, PLUGINS_PRESET_DEFAULT } from '../constants'
import {
  cleanupAttrsPlugin,
  cleanupEnableBackgroundPlugin,
  cleanupIdsPlugin,
  cleanupListOfValuesPlugin,
  cleanupNumericValuesPlugin,
  collapseGroupsPlugin,
  convertColorsPlugin,
  convertEllipseToCirclePlugin,
  convertOneStopGradientsPlugin,
  convertPathDataPlugin,
  convertShapeToPathPlugin,
  convertStyleToAttrsPlugin,
  convertTransformPlugin,
  inlineStylesPlugin,
  mergePathsPlugin,
  mergeStylesPlugin,
  minifyStylesPlugin,
  moveElemsAttrsToGroupPlugin,
  moveGroupAttrsToElemsPlugin,
  presetDefaultPlugin,
  removeAttributesBySelectorPlugin,
  removeCommentsPlugin,
  removeDeprecatedAttrsPlugin,
  removeDescPlugin,
  removeDimensionsPlugin,
  removeDoctypePlugin,
  removeEditorsNSDataPlugin,
  removeElementsByAttrPlugin,
  removeEmptyAttrsPlugin,
  removeEmptyContainersPlugin,
  removeEmptyTextPlugin,
  removeHiddenElemsPlugin,
  removeMetadataPlugin,
  removeNonInheritableGroupAttrsPlugin,
  removeTitlePlugin,
  removeUnknownsAndDefaultsPlugin,
  removeUnusedNSPlugin,
  removeUselessDefsPlugin,
  removeUselessStrokeAndFillPlugin,
  removeViewBoxPlugin,
  removeXlinkPlugin,
  removeXMLNSPlugin,
  removeXMLProcInstPlugin,
  reusePathsPlugin,
  sortAttrsPlugin,
  sortDefsChildrenPlugin,
} from './plugins'
import type { JSONSchema4 } from 'json-schema'

/**
 * Stringify Options
 *
 * Bellow options are not supported
 * - `regEntities`
 * - `regValEntities`
 * - `encodeEntity`
 */
const js2svg = {
  type: 'object',
  // @keep-sorted
  properties: {
    attrEnd: {
      type: 'string',
    },
    attrStart: {
      type: 'string',
    },
    cdataEnd: {
      type: 'string',
    },
    cdataStart: {
      type: 'string',
    },
    commentEnd: {
      type: 'string',
    },
    commentStart: {
      type: 'string',
    },
    doctypeEnd: {
      type: 'string',
    },
    doctypeStart: {
      type: 'string',
    },
    eol: {
      type: 'string',
      enum: ['lf', 'crlf'],
    },
    finalNewline: {
      type: 'boolean',
    },
    indent: {
      type: 'number',
    },
    pretty: {
      type: 'boolean',
    },
    procInstEnd: {
      type: 'string',
    },
    procInstStart: {
      type: 'string',
    },
    tagCloseEnd: {
      type: 'string',
    },
    tagCloseStart: {
      type: 'string',
    },
    tagOpenEnd: {
      type: 'string',
    },
    tagOpenStart: {
      type: 'string',
    },
    tagShortEnd: {
      type: 'string',
    },
    tagShortStart: {
      type: 'string',
    },
    textEnd: {
      type: 'string',
    },
    textStart: {
      type: 'string',
    },
    useShortTags: {
      type: 'boolean',
    },
  },
} satisfies JSONSchema4

const plugins = {
  type: 'array',
  items: {
    anyOf: [
      {
        type: 'string',
        enum: [
          PLUGIN_PRESET_DEFAULT,

          // preset-default plugin names
          ...PLUGINS_PRESET_DEFAULT,
        ],
      },
      /**
       * plugins with params
       */
      cleanupAttrsPlugin,
      cleanupEnableBackgroundPlugin,
      cleanupIdsPlugin,
      cleanupListOfValuesPlugin,
      cleanupNumericValuesPlugin,
      collapseGroupsPlugin,
      convertColorsPlugin,
      convertEllipseToCirclePlugin,
      convertOneStopGradientsPlugin,
      convertPathDataPlugin,
      convertShapeToPathPlugin,
      convertStyleToAttrsPlugin,
      convertTransformPlugin,
      inlineStylesPlugin,
      mergePathsPlugin,
      mergeStylesPlugin,
      minifyStylesPlugin,
      moveElemsAttrsToGroupPlugin,
      moveGroupAttrsToElemsPlugin,
      presetDefaultPlugin,
      removeAttributesBySelectorPlugin,
      removeCommentsPlugin,
      removeDeprecatedAttrsPlugin,
      removeDescPlugin,
      removeDimensionsPlugin,
      removeDoctypePlugin,
      removeEditorsNSDataPlugin,
      removeElementsByAttrPlugin,
      removeEmptyAttrsPlugin,
      removeEmptyContainersPlugin,
      removeEmptyTextPlugin,
      removeHiddenElemsPlugin,
      removeMetadataPlugin,
      removeNonInheritableGroupAttrsPlugin,
      removeTitlePlugin,
      removeUnknownsAndDefaultsPlugin,
      removeUselessDefsPlugin,
      removeUselessStrokeAndFillPlugin,
      removeViewBoxPlugin,
      removeXlinkPlugin,
      removeXMLNSPlugin,
      removeXMLProcInstPlugin,
      removeUnusedNSPlugin,
      reusePathsPlugin,
      sortAttrsPlugin,
      sortDefsChildrenPlugin,
    ],
  },
} satisfies JSONSchema4

export const svgoConfigProperties = {
  path: {
    type: 'string',
  },
  multipass: {
    type: 'boolean',
  },
  floatPrecision: {
    type: 'integer',
    minimum: 0,
  },
  datauri: {
    type: 'string',
    enum: ['base64', 'enc', 'unenc'],
  },
  js2svg,
  plugins,
} satisfies JSONSchema4['properties']
