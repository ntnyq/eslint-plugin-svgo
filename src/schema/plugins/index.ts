export * from './sortAttrs'
export * from './prefixIds'
export * from './cleanupIds'
export * from './mergePaths'
export * from './removeDesc'
export * from './reusePaths'
export * from './removeTitle'
export * from './mergeStyles'
export * from './removeXMLNS'
export * from './removeXlink'
export * from './removeAttrs'
export * from './inlineStyles'
export * from './cleanupAttrs'
export * from './minifyStyles'
export * from './convertColors'
export * from './presetDefault'
export * from './removeDoctype'
export * from './removeViewBox'
export * from './removeScripts'
export * from './removeComments'
export * from './removeMetadata'
export * from './collapseGroups'
export * from './removeUnusedNS'
export * from './removeEmptyText'
export * from './convertPathData'
export * from './removeDimensions'
export * from './convertTransform'
export * from './sortDefsChildren'
export * from './removeEmptyAttrs'
export * from './removeHiddenElems'
export * from './removeUselessDefs'
export * from './removeXMLProcInst'
export * from './removeStyleElement'
export * from './removeRasterImages'
export * from './convertShapeToPath'
export * from './removeEditorsNSData'
export * from './convertStyleToAttrs'
export * from './removeEditorsNSData'
export * from './cleanupListOfValues'
export * from './removeElementsByAttr'
export * from './cleanupNumericValues'
export * from './removeOffCanvasPaths'
export * from './moveElemsAttrsToGroup'
export * from './moveGroupAttrsToElems'
export * from './removeDeprecatedAttrs'
export * from './removeEmptyContainers'
export * from './convertEllipseToCircle'
export * from './addClassesToSVGElement'
export * from './convertOneStopGradients'
export * from './cleanupEnableBackground'
export * from './removeUnknownsAndDefaults'
export * from './addAttributesToSVGElement'
export * from './removeUselessStrokeAndFill'
export * from './removeAttributesBySelector'
export * from './removeNonInheritableGroupAttrs'
