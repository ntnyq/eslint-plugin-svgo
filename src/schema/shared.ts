import type { JSONSchema4 } from 'json-schema'

export const booleanSchema: JSONSchema4 = {
  type: 'boolean',
}

export const numberSchema: JSONSchema4 = {
  type: 'number',
}

export const onlyFalseSchema: JSONSchema4 = {
  type: 'boolean',
  enum: [false],
}

export const stringSchema: JSONSchema4 = {
  type: 'string',
}

export const integerSchema: JSONSchema4 = {
  type: 'integer',
}

export const objectSchema: JSONSchema4 = {
  type: 'object',
  additionalProperties: false,
}

export const stringArraySchema: JSONSchema4 = {
  type: 'array',
  items: stringSchema,
}

export const precisionSchema: JSONSchema4 = {
  ...integerSchema,
  minimum: 0,
}

// Common pattern: boolean or string union schema
export const booleanOrStringSchema: JSONSchema4 = {
  oneOf: [booleanSchema, stringSchema],
}

// Common pattern: string or string array union schema
export const stringOrStringArraySchema: JSONSchema4 = {
  oneOf: [stringSchema, stringArraySchema],
}

// Common pattern: positive number schema
export const positiveNumberSchema: JSONSchema4 = {
  ...numberSchema,
  minimum: 0,
}

// Common enums
export const COLOR_CASE_VALUES = ['lower', 'upper']
export const EOL_VALUES = ['lf', 'crlf']
export const DATAURI_VALUES = ['base64', 'enc', 'unenc']
export const XMLNS_ORDER_VALUES = ['front', 'alphabetical']

// Common enum schemas
export const colorCaseSchema: JSONSchema4 = {
  ...stringSchema,
  enum: COLOR_CASE_VALUES,
}

export const eolSchema: JSONSchema4 = {
  ...stringSchema,
  enum: EOL_VALUES,
}

export const datauriSchema: JSONSchema4 = {
  ...stringSchema,
  enum: DATAURI_VALUES,
}

export const xmlnsOrderSchema: JSONSchema4 = {
  ...stringSchema,
  enum: XMLNS_ORDER_VALUES,
}
