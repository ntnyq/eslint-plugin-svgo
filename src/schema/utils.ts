import type { JSONSchema4 } from 'json-schema'

export function createParamsSchema(
  properties: JSONSchema4['properties'] = {},
  additionalProperties = false,
) {
  const schema: JSONSchema4 = {
    type: 'object',
    properties,
    additionalProperties,
  }
  return schema
}

export function createPluginSchema(pluginName: string, params?: JSONSchema4) {
  const schema: JSONSchema4 = {
    type: 'object',
    properties: {
      name: {
        type: 'string',
        enum: [pluginName],
      },
      ...(params && Object.keys(params.properties || {}).length > 0
        ? { params }
        : {}),
    },
    required: ['name'],
    additionalProperties: false,
  }

  return schema
}
