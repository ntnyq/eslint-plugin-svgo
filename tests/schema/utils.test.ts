import { describe, expect, it } from 'vitest'
import { createParamsSchema, createPluginSchema } from '../../src/schema/utils'
import type { JSONSchema4 } from 'json-schema'

describe('createParamsSchema', () => {
  it('should create empty params schema with default values', () => {
    const schema = createParamsSchema()

    expect(schema).toEqual({
      type: 'object',
      properties: {},
      additionalProperties: false,
    })
  })

  it('should create params schema with provided properties', () => {
    const properties: JSONSchema4['properties'] = {
      enabled: { type: 'boolean' },
      value: { type: 'string' },
    }
    const schema = createParamsSchema(properties)

    expect(schema).toEqual({
      type: 'object',
      properties,
      additionalProperties: false,
    })
  })

  it('should create params schema with additionalProperties enabled', () => {
    const properties: JSONSchema4['properties'] = {
      test: { type: 'boolean' },
    }
    const schema = createParamsSchema(properties, true)

    expect(schema).toEqual({
      type: 'object',
      properties,
      additionalProperties: true,
    })
  })
})

describe('createPluginSchema', () => {
  it('should create plugin schema without params', () => {
    const schema = createPluginSchema('testPlugin')

    expect(schema).toEqual({
      type: 'object',
      properties: {
        name: {
          type: 'string',
          enum: ['testPlugin'],
        },
      },
      required: ['name'],
      additionalProperties: false,
    })
  })

  it('should create plugin schema with valid params', () => {
    const params: JSONSchema4 = {
      type: 'object',
      properties: {
        enabled: { type: 'boolean' },
        value: { type: 'string' },
      },
      additionalProperties: false,
    }

    const schema = createPluginSchema('testPlugin', params)

    expect(schema).toEqual({
      type: 'object',
      properties: {
        name: {
          type: 'string',
          enum: ['testPlugin'],
        },
        params,
      },
      required: ['name'],
      additionalProperties: false,
    })
  })

  it('should create plugin schema without params when params has no properties', () => {
    const params: JSONSchema4 = {
      type: 'object',
      properties: {},
      additionalProperties: false,
    }

    const schema = createPluginSchema('testPlugin', params)

    expect(schema).toEqual({
      type: 'object',
      properties: {
        name: {
          type: 'string',
          enum: ['testPlugin'],
        },
      },
      required: ['name'],
      additionalProperties: false,
    })
  })

  it('should create plugin schema without params when params is undefined', () => {
    const schema = createPluginSchema('testPlugin', undefined)

    expect(schema).toEqual({
      type: 'object',
      properties: {
        name: {
          type: 'string',
          enum: ['testPlugin'],
        },
      },
      required: ['name'],
      additionalProperties: false,
    })
  })

  it('should create plugin schema without params when params has undefined properties', () => {
    const params: JSONSchema4 = {
      type: 'object',
      additionalProperties: false,
    }

    const schema = createPluginSchema('testPlugin', params)

    expect(schema).toEqual({
      type: 'object',
      properties: {
        name: {
          type: 'string',
          enum: ['testPlugin'],
        },
      },
      required: ['name'],
      additionalProperties: false,
    })
  })

  it('should handle complex plugin names with special characters', () => {
    const pluginName = 'preset-default'
    const schema = createPluginSchema(pluginName)

    expect(schema.properties?.name).toEqual({
      type: 'string',
      enum: [pluginName],
    })
  })

  it('should handle complex params schema structure', () => {
    const params: JSONSchema4 = {
      type: 'object',
      properties: {
        floatPrecision: {
          type: 'integer',
          minimum: 0,
        },
        overrides: {
          type: 'object',
          properties: {
            cleanupIds: {
              oneOf: [
                { type: 'boolean' },
                {
                  type: 'object',
                  properties: {
                    minify: { type: 'boolean' },
                  },
                },
              ],
            },
          },
          additionalProperties: false,
        },
      },
      additionalProperties: false,
    }

    const schema = createPluginSchema('preset-default', params)

    expect(schema).toEqual({
      type: 'object',
      properties: {
        name: {
          type: 'string',
          enum: ['preset-default'],
        },
        params,
      },
      required: ['name'],
      additionalProperties: false,
    })
  })

  describe('edge cases', () => {
    it('should handle empty string plugin name', () => {
      const schema = createPluginSchema('')

      expect(schema.properties?.name).toEqual({
        type: 'string',
        enum: [''],
      })
    })

    it('should handle plugin name with spaces', () => {
      const pluginName = 'my plugin name'
      const schema = createPluginSchema(pluginName)

      expect(schema.properties?.name).toEqual({
        type: 'string',
        enum: [pluginName],
      })
    })

    it('should include params when params has at least one property', () => {
      const params: JSONSchema4 = {
        type: 'object',
        properties: {
          singleProp: { type: 'boolean' },
        },
        additionalProperties: false,
      }

      const schema = createPluginSchema('testPlugin', params)

      expect(schema.properties).toHaveProperty('params')
      expect(schema.properties?.params).toBe(params)
    })

    it('should exclude params when params object exists but properties is empty', () => {
      const params: JSONSchema4 = {
        type: 'object',
        properties: {},
        additionalProperties: false,
      }

      const schema = createPluginSchema('testPlugin', params)

      expect(schema.properties).not.toHaveProperty('params')
    })
  })

  describe('real-world examples', () => {
    it('should work with cleanupIds plugin schema', () => {
      const params: JSONSchema4 = {
        type: 'object',
        properties: {
          remove: { type: 'boolean' },
          minify: { type: 'boolean' },
          force: { type: 'boolean' },
          preserve: {
            type: 'array',
            items: { type: 'string' },
          },
          preservePrefixes: {
            type: 'array',
            items: { type: 'string' },
          },
        },
        additionalProperties: false,
      }

      const schema = createPluginSchema('cleanupIds', params)

      expect(schema).toMatchSnapshot()
    })

    it('should work with removeMetadata plugin (no params)', () => {
      const schema = createPluginSchema('removeMetadata')

      expect(schema).toMatchSnapshot()
    })

    it('should work with prefixIds plugin schema', () => {
      const params: JSONSchema4 = {
        type: 'object',
        properties: {
          delim: { type: 'string' },
          prefix: {
            oneOf: [{ type: 'boolean' }, { type: 'string' }],
          },
          prefixIds: { type: 'boolean' },
          prefixClassNames: { type: 'boolean' },
        },
        additionalProperties: false,
      }

      const schema = createPluginSchema('prefixIds', params)

      expect(schema).toMatchSnapshot()
    })
  })
})
