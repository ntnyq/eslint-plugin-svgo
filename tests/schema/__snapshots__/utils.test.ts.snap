// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createPluginSchema > real-world examples > should work with cleanupIds plugin schema 1`] = `
{
  "additionalProperties": false,
  "properties": {
    "name": {
      "enum": [
        "cleanupIds",
      ],
      "type": "string",
    },
    "params": {
      "additionalProperties": false,
      "properties": {
        "force": {
          "type": "boolean",
        },
        "minify": {
          "type": "boolean",
        },
        "preserve": {
          "items": {
            "type": "string",
          },
          "type": "array",
        },
        "preservePrefixes": {
          "items": {
            "type": "string",
          },
          "type": "array",
        },
        "remove": {
          "type": "boolean",
        },
      },
      "type": "object",
    },
  },
  "required": [
    "name",
  ],
  "type": "object",
}
`;

exports[`createPluginSchema > real-world examples > should work with prefixIds plugin schema 1`] = `
{
  "additionalProperties": false,
  "properties": {
    "name": {
      "enum": [
        "prefixIds",
      ],
      "type": "string",
    },
    "params": {
      "additionalProperties": false,
      "properties": {
        "delim": {
          "type": "string",
        },
        "prefix": {
          "oneOf": [
            {
              "type": "boolean",
            },
            {
              "type": "string",
            },
          ],
        },
        "prefixClassNames": {
          "type": "boolean",
        },
        "prefixIds": {
          "type": "boolean",
        },
      },
      "type": "object",
    },
  },
  "required": [
    "name",
  ],
  "type": "object",
}
`;

exports[`createPluginSchema > real-world examples > should work with removeMetadata plugin (no params) 1`] = `
{
  "additionalProperties": false,
  "properties": {
    "name": {
      "enum": [
        "removeMetadata",
      ],
      "type": "string",
    },
  },
  "required": [
    "name",
  ],
  "type": "object",
}
`;
